Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -114,65 +116,111 @@ ISymbolTableRelocator symbolTableRelocator
 
       databaseEntry.CancelPendingUpdates();
       var cancellationSource = new CancellationTokenSource();
+      var previousDocumentTask = FirstSuccessfulAsync(databaseEntry.VerifiedDocument, databaseEntry.ResolvedDocument);
+      var resolvedDocumentTask = ApplyChangesAsync(previousDocumentTask, documentChange, cancellationSource.Token);
+      var verifiedDocument = VerifyAsync(resolvedDocumentTask, VerifyOnChange, cancellationSource.Token);
       var updatedEntry = new DocumentEntry(
         documentChange.TextDocument.Version,
-        Task.Run(async () => await ApplyChangesAsync(await databaseEntry.Document, documentChange, cancellationSource.Token)),
+        resolvedDocumentTask,
+        verifiedDocument,
         cancellationSource
       );
       documents[documentUri] = updatedEntry;
-      return await updatedEntry.Document;
+      return resolvedDocumentTask.ToObservable().Concat(verifiedDocument.ToObservable());
+    }
+
+
+    /// <summary>
+    /// Returns the first task that completes successfully.
+    /// If multiple tasks are already completed, the first one is returned.
+    /// </summary>
+    private static Task<T> FirstSuccessfulAsync<T>(params Task<T>[] tasks) {
```

Comment:
Maybe it makes sense to create a separate class for this method so it can be unit tested.

Reply:
Thanks! I added unit tests and they helped spot a rare concurrency bug that I'm guessing has been the reason the build was failing non-deterministically.